### PR TITLE
Preserve data in json files

### PR DESCRIPTION
**Pull request by ismayito**
### Preserve data in files in ruby
_______________________________________________________________________________________________________
#### Summary
_______________________________________________________________________________________________________
In this project, we added the following:
- preserved data into JSON files using ruby
- Loaded data from  json files on app load  using ruby
- Fix linter errors
- Followed gitflow

### CODE REVIEW

**Review by DammyShittu — COMMENTED** [review]
Hi @ismayito and @rmiyoyo,

While you made a great effort in this project, I cannot proceed to review your code.

## Invalid Code Review Request

### The second team member, @rmiyoyo, did not make a meaningful contribution to this project. He mostly only added comments which, while important, are relatively not as important as actual code.

### Also ensure rentals can be created seamlessly without the app crashing or having to quit the app and try again. All options should work at a go.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by rmiyoyo** [comment]
# Work Done
- Fixed existing bugs and resubmitted for review.
- Data is saved; one per type of data stored and named accordingly.
- On startup the app loads all data from files.

File: create_book.rb [hunk 1]
```diff
@@ -12,11 +14,40 @@ def create_book
...
     book_author = gets.chomp
+
+    # Load existing books from the JSON file
+    existing_books = load_books_from_json
+
     added_book = Book.new(book_name, book_author)
     @books.push(added_book)
+    existing_books.push({ title: added_book.title, author: added_book.author })
+
+    save_books_to_json(existing_books)
```
Comment (zeff96) [inline create_book.rb:25]: - [ ] Hey Team! 👋 Stellar job on implementing the method for saving books to JSON. 📚 However, there's a slight deviation from the project requirement. 🚧 According to the specifications, we're supposed to save to JSON only on app exit, not immediately upon book creation. 🔄 Kindly review the current implementation and make the necessary adjustments to align with the project's directive. Let's ensure we save that data only when the app bids us farewell. 🌟 Thanks for your attention to detail! 🙌

File: create_rental.rb [hunk 2]
```diff
@@ -2,20 +2,26 @@
...
-  book = @books.find { |b| b.title == book_title }
+  book = @books.find { |b| b['title'] == book_title }
+  book_hashed = Book.new(book['title'], book['author'])
   print 'Enter rental date: '
   rental_date = gets.chomp
-  rental = Rental.new(rental_date, person, book)
+  rental = Rental.new(rental_date, person_hashed, book_hashed)
   @rentals << rental
+  # method to rentals student to 'books.json'
+  save_rentals_to_json(@rentals)
```
Comment (zeff96) [inline create_rental.rb:24]: - [ ] Hey Team! 👋 Stellar job on implementing the method for saving rentals to JSON. 📚 However, there's a slight deviation from the project requirement. 🚧 According to the specifications, we're supposed to save to JSON only on app exit, not immediately upon rental creation. 🔄 Kindly review the current implementation and make the necessary adjustments to align with the project's directive. Let's ensure we save that data only when the app bids us farewell. 🌟 Thanks for your attention to detail! 🙌

File: create_student.rb [hunk 3]
```diff
@@ -7,6 +8,11 @@ def initialize(person, app_instance)
     @app_instance = app_instance
   end
 
+  # method to save student to 'books.json'
+  def save_persons_to_json(persons)
+    File.write('person.json', JSON.pretty_generate(persons.map(&:to_hash)))
+  end
```
Comment (zeff96) [inline create_student.rb:14]: - [ ] Hey Team! 👋 Stellar job on implementing the method for saving person to JSON. 📚 However, there's a slight deviation from the project requirement. 🚧 According to the specifications, we're supposed to save to JSON only on app exit, not immediately upon person creation. 🔄 Kindly review the current implementation and make the necessary adjustments to align with the project's directive. Let's ensure we save that data only when the app bids us farewell. 🌟 Thanks for your attention to detail! 🙌

File: create_teacher.rb [hunk 4]
```diff
@@ -7,6 +8,11 @@ def initialize(person, app_instance)
     @app_instance = app_instance
   end
 
+  # method to save the tearcher into person.json file
+  def save_persons_to_json(persons)
+    File.write('person.json', JSON.pretty_generate(persons.map(&:to_hash)))
+  end
```
Comment (zeff96) [inline create_teacher.rb:14]: - [ ] Hey Team! 👋 Stellar job on implementing the method for saving person to JSON. 📚 However, there's a slight deviation from the project requirement. 🚧 According to the specifications, we're supposed to save to JSON only on app exit, not immediately upon person creation. 🔄 Kindly review the current implementation and make the necessary adjustments to align with the project's directive. Let's ensure we save that data only when the app bids us farewell. 🌟 Thanks for your attention to detail! 🙌

**Review by zeff96 — CHANGES_REQUESTED** [review]
### Status: CHANGES REQUIRED ♻️
Hi @ismayito and @rmiyoyo , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
:heavy_check_mark: Linters passing
:heavy_check_mark: Descriptive PR



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by rmiyoyo** [comment]
# Work Done After Review
- Data is now saved when user chooses to exit application

**Review by binhussen — APPROVED** [review]
Hi @Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉
IApproveSimonCowellGIF (6)

## Highlights
- Preserved Data to Json file ✔️
- Well Starctured Code ✔️
- Used Git flow ✔️
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**